test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpWriteContext.java

```diff
@@ -0,0 +1,157 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio.channel;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.transport.nio.SocketSelector;
+import org.elasticsearch.transport.nio.WriteOperation;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.util.LinkedList;
+
+public class TcpWriteContext implements WriteContext {
+
+    private static final ClosedChannelException CLOSED_CHANNEL_EXCEPTION = new ClosedChannelException();
+
+    private final NioSocketChannel channel;
+    private final LinkedList<WriteOperation> queued = new LinkedList<>();
+
+    public TcpWriteContext(NioSocketChannel channel) {
+        this.channel = channel;
+    }
+
+    @Override
+    public void sendMessage(BytesReference reference, ActionListener<NioChannel> listener) {
+        if (channel.isWritable() == false) {
+            listener.onFailure(CLOSED_CHANNEL_EXCEPTION);
+            return;
+        }
+
+        WriteOperation writeOperation = new WriteOperation(channel, reference, listener);
+        SocketSelector selector = channel.getSelector();
+        if (selector.isOnCurrentThread() == false) {
+            selector.queueWrite(writeOperation);
+            return;
+        }
+
+        // TODO: Eval if we will allow writes from sendMessage
+//        if (queued.isEmpty() == false) {
+        selector.queueWriteInChannelBuffer(writeOperation);
+//        } else {
+//            singleFlush(writeOperation);
+//        }
+    }
+
+    @Override
+    public void queueWriteOperations(WriteOperation writeOperation) {
+        assert channel.getSelector().isOnCurrentThread();
+        queued.add(writeOperation);
+    }
+
+    @Override
+    public void flushChannel() throws IOException {
+        assert channel.getSelector().isOnCurrentThread();
+        if (queued.size() == 1) {
+            singleFlush(queued.pop());
+        } else {
+            multiFlush();
+        }
+    }
+
+    @Override
+    public boolean hasQueuedWriteOps() {
+        assert channel.getSelector().isOnCurrentThread();
+        return queued.isEmpty() == false;
+    }
+
+    @Override
+    public void clearQueuedWriteOps(Exception e) {
+        assert channel.getSelector().isOnCurrentThread();
+        for (WriteOperation op : queued) {
+            op.getListener().onFailure(e);
+        }
+        queued.clear();
+    }
+
+    private void singleFlush(WriteOperation headOp) throws IOException {
+        ByteBuffer[] buffers = headOp.getBuffers();
+        long remaining = headOp.bytesRemaining();
+        long written;
+        if (buffers.length == 1) {
+            written = channel.write(buffers[0]);
+        } else {
+            written = channel.vectorizedWrite(buffers);
+        }
+
+        if (written < remaining) {
```

Comment:
The ByteBuffers have internal markers that the low-level flush operations use. So that is why stuff is not written twice. The WriteOperation#decrementRemaining work is so we know when the total WriteOperation (multiple buffers) is complete and can call the listener. 

I don't disagree with your thought about moving writing into WriteOperation though. I'm taking a look at what that might look like.